### PR TITLE
feat(optimize): support MPS position exposure repair

### PR DESCRIPTION
**Pull request by enarjord**
Summary:
- model Trailing Martingale per-position exposure repair in the single- and multi-coin Metal proxies for long, short, dual-side, and suite-compatible runs
- preserve exact Rust precedence, passive pricing, strict below-target sizing, tunable thresholds, and static per-coin enable/threshold overrides
- keep EMA Anchor position repair and side-wide total-exposure repair fail closed; document the supported boundary

Validation:
- 262 Rust tests passed with no default features
- cargo check with tests and cargo fmt check passed
- full Python suite passed
- 72 real Apple MPS tests passed on this M3 MacBook Air
- focused single/multi, long/short, global/per-coin reducer regressions passed on real MPS
- extension source fingerprint verified after rebuild
- end-to-end ETH long-only MPS optimization over 68 days completed: 1,024 proxy evaluations, 16 exact Rust validations, 8 generations, no optimizer halt; Pareto configs persisted enabled repair with optimized thresholds

Scope:
- GPU-only additive path; CPU optimization, backtesting, and bot operation are unchanged
- exact Rust results remain authoritative through existing feasibility, rank, and drift gates

### CODE REVIEW

**Comment by enarjord** [comment]
@codex review

**Comment by enarjord** [comment]
Addressed both exact-head findings in 0be01d086:

- directional TM now preserves reducer reachability through grid pruning, executes the sorted-nearest WEL reducer first, then reconstructs ordinary closes from the post-repair size
- multicoin TM now reconstructs the bounded post-reducer recursive grid and retains all next-candle-reachable quantity/notional instead of dropping ordinary closes
- added real-MPS long/short regressions for both cases

Validation on this head:
- cargo test --no-default-features: 262 passed
- cargo check --tests; cargo check --no-default-features; cargo fmt --check: passed
- full Python suite: passed
- full real Apple MPS suite: 76 passed
- end-to-end ETH local-cache GPU opt: 1,024 proxy evals, 32 exact Rust validations, 8 generations, no halt

@codex review

File: passivbot-rust/src/gpu/mps_trailing_martingale_directional.metal [hunk 1]
```diff
@@ -358,6 +362,43 @@ inline void generate_orders(
...
+            if (new_we >= wel_target - 1.0e-12f && reducer_qty < s.psize) {
+                reducer_qty = fmin(
+                    s.psize,
+                    fmax(reducer_min, ceil_step(reducer_qty + qty_step, qty_step))
+                );
+            }
+            s.close_ticks = reducer_ticks;
+            s.close_price = reducer_price;
+            s.close_qty = reducer_qty;
+            return;
```
[the remark this answers is not in the document]
Comment (enarjord) [reply passivbot-rust/src/gpu/mps_trailing_martingale_directional.metal:405]: Fixed in 0be01d086: reducer reachability is retained through the conservative grid scan, the reducer is executed first in sorted price order, and the remaining grid is rebuilt from the post-repair size. Covered by real-MPS long/short regression tests.

File: passivbot-rust/src/gpu/mps_trailing_martingale_multicoin.metal [hunk 2]
```diff
@@ -913,6 +922,48 @@ inline void passivbot_trailing_martingale_multicoin_impl(
...
+                                psize[c],
+                                fmax(
+                                    reducer_min,
+                                    ceil_step(reducer_qty + qty_step, qty_step)
+                                )
+                            );
+                        }
+                        close_qty[c] = reducer_qty;
+                        close_tick[c] = reducer_tick;
+                        continue;
```
[the remark this answers is not in the document]
Comment (enarjord) [reply passivbot-rust/src/gpu/mps_trailing_martingale_multicoin.metal:1255]: Fixed in 0be01d086: multicoin grid mode reconstructs all bounded post-reducer rungs and aggregates every next-candle-reachable quantity/notional. Covered for long and short on real Apple MPS.

File: passivbot-rust/src/gpu/mps_trailing_martingale_multicoin.metal [hunk 3]
```diff
@@ -92,6 +92,123 @@ inline float calc_close_qty(
...
+                short_side ? 1.0f - threshold : 1.0f + threshold
+            );
+            int target_tick = short_side
+                ? int(floor(target / price_step + 1.0e-6f))
+                : int(ceil(target / price_step - 1.0e-6f));
+            int close_touch = short_side ? touch_down : touch_up;
+            bool touch_controls = (short_side && threshold <= 0.0f)
+                || (short_side
+                    ? close_touch < target_tick
+                    : close_touch > target_tick);
```
[the remark this answers is not in the document]
Comment (enarjord) [reply passivbot-rust/src/gpu/mps_trailing_martingale_multicoin.metal]: Fixed in bd411344d: post-reducer short grid reconstruction now selects the touch only when bid < target, matching min(bid, target). Added a real-MPS case where the reducer at 102 fills but the negative-threshold target at 101 remains below the next 101.5 low.

File: passivbot-rust/src/gpu/mps_trailing_martingale_multicoin.metal [hunk 4]
```diff
@@ -375,8 +499,10 @@ inline void passivbot_trailing_martingale_multicoin_impl(
                     }
                     position_open_k[c] = -1.0f;
                 }
-                day_volume += fabs(adjusted) * fill_price * c_mult / balance;
+                day_volume += adjusted_notional * c_mult / balance;
```
[the remark this answers is not in the document]
Comment (enarjord) [reply passivbot-rust/src/gpu/mps_trailing_martingale_multicoin.metal]: Fixed in bd411344d: removed quantity/notional aggregation. The kernel reconstructs immutable generated groups at fill time and applies each reachable reducer/grid fill sequentially, updating balance before each volume contribution. Added exact numeric long/short real-MPS regressions.

**Comment by enarjord** [comment]
Second review round addressed in bd411344d:

- short post-repair grid uses exact min(bid, target) semantics for zero/negative thresholds
- removed aggregate fill normalization; reachable reducer/grid rungs now execute sequentially against each post-fill balance
- added real-MPS negative-short-target and numeric per-fill-volume regressions

Validation on this exact head:
- cargo test --no-default-features: 262 passed
- cargo check --tests; cargo check --no-default-features; cargo fmt --check: passed
- full Python suite: passed
- full real Apple MPS suite: 79 passed
- end-to-end ETH local-cache GPU opt: 1,024 proxy evals, 32 exact Rust validations, 8 generations, no halt

@codex review

File: passivbot-rust/src/gpu/mps_trailing_martingale_multicoin.metal [hunk 5]
```diff
@@ -365,18 +490,115 @@ inline void passivbot_trailing_martingale_multicoin_impl(
                     ? pprice[c] - fill_price
                     : fill_price - pprice[c]);
                 balance += pnl - adjusted * fill_price * c_mult * maker_fee;
-                float new_size = fmax(round_step(psize[c] - adjusted, qty_step), 0.0f);
-                bool went_flat = new_size <= 0.0f;
-                psize[c] = new_size;
+                psize[c] = fmax(
+                    round_step(psize[c] - adjusted, qty_step), 0.0f
+                );
+                day_volume += adjusted * fill_price * c_mult / balance;
```
[the remark this answers is not in the document]
Comment (enarjord) [reply passivbot-rust/src/gpu/mps_trailing_martingale_multicoin.metal]: Fixed in 2063c1c0d. Both directional and multicoin TM paths now reconstruct the post-repair ladder first, merge the reducer quantity into the first executable group when both share a price tick, and perform fee/balance/volume accounting as one combined fill. Distinct-tick closes remain sequential. Added real-MPS long/short regressions for both paths with nonzero maker fees.

**Comment by enarjord** [comment]
Addressed the latest exact-head review finding in 2063c1c0d.

Validation on Apple M3/MPS:
- 83/83 MPS optimizer tests
- full Python suite
- 262 Rust tests (`--no-default-features`)
- `cargo check --all-targets`
- `cargo fmt --all -- --check`
- local-cache ETH GPU smoke: 1,024 proxy evaluations, 32 exact validations, 8 generations, no halt

@codex review

**Comment by enarjord** [comment]
Addressed both latest exact-head findings in f8c16132e:

- reducer and rebuilt-grid reachability are independent in directional and multicoin TM
- strict WEL repair retains sub-float-epsilon excesses and iterates against rounded stored size
- added eight real-MPS boundary regressions across long/short and single/multicoin

Validation on Apple M3/MPS:
- 91/91 MPS optimizer tests
- full Python suite
- 262 Rust tests (`--no-default-features`)
- `cargo check --all-targets`
- `cargo fmt --all -- --check`
- local-cache ETH GPU smoke: 1,024 proxy evaluations, 32 exact validations, 8 generations, no halt

@codex review

File: passivbot-rust/src/gpu/mps_trailing_martingale_multicoin.metal [hunk 6]
```diff
@@ -360,23 +485,173 @@ inline void passivbot_trailing_martingale_multicoin_impl(
...
+                float coin_close_threshold_we = 0.0f;
+                float coin_close_threshold_v1h = 0.0f;
+                float coin_close_threshold_v1m = 0.0f;
+                CloseGroup group;
+                int group_count = 0;
+                bool reverse = false;
+                bool merge_reducer_with_first = false;
+
+                if (close_reconstruct_after_reducer[c]
+                    && grid_gen_psize > 0.0f) {
```
[the remark this answers is not in the document]
Comment (enarjord) [reply passivbot-rust/src/gpu/mps_trailing_martingale_multicoin.metal]: Fixed in f8c16132e. Directional TM now conservatively scans the nearest touch tick and checks reducer reachability separately from every reconstructed group. Multicoin reconstructs whenever a reducer-backed grid exists, independently tests reducer and group ticks, and records fill state only if either executes. Added real-MPS long/short tests where the nearest grid fills one tick before the directional reducer in both single- and multi-coin kernels.

File: passivbot-rust/src/gpu/mps_trailing_martingale_directional.metal [hunk 7]
```diff
@@ -358,6 +365,44 @@ inline void generate_orders(
...
+        int reducer_ticks = is_long ? touch_up_ticks : touch_down_ticks;
+        float reducer_price = float(reducer_ticks) * price_step;
+        if (reducer_ticks > 0 && reducer_price > 0.0f) {
+            float reducer_min = min_entry_qty(
+                reducer_price, qty_step, min_qty, min_cost, c_mult
+            );
+            float target_psize = wel_target * balance
+                / fmax(s.pprice * c_mult, 1.0e-12f);
+            float reduce_qty = fmax(s.psize - target_psize, 0.0f);
+            if (reduce_qty <= 1.1920928955078125e-7f) reduce_qty = qty_step;
```
[the remark this answers is not in the document]
Comment (enarjord) [reply passivbot-rust/src/gpu/mps_trailing_martingale_directional.metal]: Fixed in f8c16132e. Both kernels retain positive excesses above f64 epsilon, iterate with the same 10,000-step safeguard as Rust, and compare the rounded post-execution position size against the strict target so storage quantization cannot leave exposure on or above it. Added real-MPS long/short sub-unit, 1e-8-step regressions for directional and multicoin paths.

File: passivbot-rust/src/gpu/mps_trailing_martingale_directional.metal [hunk 8]
```diff
@@ -635,23 +691,68 @@ inline void passivbot_single_coin_impl(
...
+            bool merge_reducer_with_first = false;
+            if (reducer_reachable && group_count > 0) {
+                int first_wanted = reverse ? group_count - 1 : 0;
+                recursive_close_groups(
+                    grid_source, true, first_wanted, qty_step, price_step,
+                    min_qty, min_cost, c_mult, group
+                );
+                merge_reducer_with_first = group.ticks == reducer_ticks;
+            }
+            if (reducer_reachable && !merge_reducer_with_first) {
```
[the remark this answers is not in the document]
Comment (enarjord) [reply passivbot-rust/src/gpu/mps_trailing_martingale_directional.metal]: Fixed in c25594f1c. Directional and multicoin execution now insert the reachable reducer into the ordinary close groups by Rust price order (long ascending, short descending), merge only equal ticks, and preserve the per-fill post-balance volume denominator. Added real-MPS long/short single- and multicoin off-tick ordering regressions with explicit balance and volume expectations; all 8 new edge cases and the 31-case reducer matrix pass.

File: passivbot-rust/src/gpu/mps_trailing_martingale_multicoin.metal [hunk 9]
```diff
@@ -92,6 +92,121 @@ inline float calc_close_qty(
...
+    selected.ticks = 0;
+    selected.price = 0.0f;
+    selected.qty = 0.0f;
+    float sim_psize = psize;
+    bool have_group = false;
+    int group_count = 0;
+    int group_ticks = 0;
+    float group_price = 0.0f;
+    float group_qty = 0.0f;
+    for (int rung = 0; rung < 500 && sim_psize > 0.0f; ++rung) {
```
[the remark this answers is not in the document]
Comment (enarjord) [reply passivbot-rust/src/gpu/mps_trailing_martingale_multicoin.metal]: Fixed in c25594f1c. A generated reducer now consumes the first shared close-generation slot: directional reconstruction uses 499 rungs only for reducer-backed grids, and multicoin post-reducer reconstruction is capped at 499. Added real-MPS long/short single- and multicoin regressions whose residual size distinguishes 499 from 500 grid closes.

**Comment by enarjord** [comment]
Review fixes are on exact head c25594f1c.\n\nAddressed both findings:\n- reachable reducer and rebuilt grid close groups now execute in exact Rust price order for long and short, with equal-tick merging and per-fill balance denominators;\n- reducer-backed recursive grids now use the remaining 499 slots from the shared 500-order cap.\n\nValidation on this head:\n- 8/8 new real-MPS edge cases;\n- 31/31 focused real-MPS reducer/grid matrix;\n- complete real-MPS optimizer test file passed;\n- full Python suite passed after temporarily isolating and byte-for-byte restoring the two known generated local symbol maps;\n- cargo test --no-default-features: 262 passed;\n- cargo check --all-targets passed;\n- cargo fmt --all -- --check passed;\n- ETH local-cache smoke: 1,024 proxy evaluations, 32 exact validations, 8 generations, 5.9s, no drift or constraint halt.\n\nAll review threads are resolved. @codex review
